faw/pdf-observatory/faw_analysis_set_parse.py

```diff
@@ -327,56 +338,89 @@ def get_cfg(k):
         for k in parser_set:
             # Check version info -- not equal means it needs to run; equal means
             # it needs to be marked done
+            cfg = all_parser_configs.get(k, {})
             ver_db_info = versions_keyed.get(k, [[None, None], [], []])
             ver_db = ver_db_info[0]
             ver_cfg = doc['parsers'][k]
 
-            cfg = get_cfg(k)
-            if cfg.get('disabled'):
+            if cfg.get('disabled', False):
                 # We *do* want to unset based on version. A different version
                 # may not be disabled.
                 parsers_done[k] = ver_cfg
                 continue
 
-            if ver_db[0] != ver_cfg[0]:
-                parsers_to_run_tool[k] = ver_cfg[0]
-                parsers_to_run_parser[k] = ver_cfg[1]
-
-                # Delete matching tool documents
-                db['rawinvocations'].delete_many({'_id': {'$in': [
-                        v['_id'] for v in ver_db_info[1]]}})
-                db['invocationsparsed'].delete_many({'_id': {'$in': [
-                        v['_id'] for v in ver_db_info[2]]}})
-            elif ver_db[1] != ver_cfg[1]:
-                parsers_to_run_parser[k] = ver_cfg[1]
+            if ver_db[0] != ver_cfg[0]:  # tool versions
+                parsers_to_run_tool.add(k)
+                artifact_dependency_graph.add_parsers_downstream_from_parser(k, parsers_to_run_tool)
+                # We'll delete matching documents and add to parsers_to_run_parser after this loop
+            elif ver_db[1] != ver_cfg[1]:  # parse versions
+                parsers_to_run_parser.add(k)
                 db['invocationsparsed'].delete_many({'_id': {'$in': [
                         v['_id'] for v in ver_db_info[2]]}})
 
             # Regardless, need to set this as done when we've run everything
             parsers_done[k] = ver_cfg
 
+        parsers_to_run_parser |= parsers_to_run_tool
+        parsers_upstream = artifact_dependency_graph.parsers_upstream_from_parsers(
+            parsers_to_run_tool
+        ) - parsers_to_run_tool
+
+        parse_versions = {}
+        for k in parsers_to_run_parser | parsers_upstream:
+            if k in doc['parsers']:
+                parse_versions[k] = doc['parsers'][k][1]
+            else:  # version didn't change; use latest parse version from config
```

Comment:
I've tested this thoroughly, but let me know if there's a case where pulling the version from the config like this will cause issues.

Reply:
This is no longer necessary now that the upstream parser versions are all in `as_parse`.